File: backend/src/api/feedback.ts
/**
 * /api/feedback — accepts user feedback on an assistant message and pushes it
 * to Langfuse as a Score attached to the trace (and, optionally, the specific
 * turn observation that produced the message).
 *
 * Body:
 *   {
 *     trace_id:        string   // from the SSE 'trace' event
 *     observation_id?: string   // turn span id (optional but recommended)
 *     value:           1 | -1 | number   // thumbs-up = 1, thumbs-down = -1
 *     comment?:        string
 *     name?:           string   // defaults to 'user-feedback'
 *     household_id?:   string   // forwarded into Langfuse metadata
 *     chat_id?:        string
 *     turn?:           number
 *   }
 *
 * Response: { ok: true } once the score is queued. The Langfuse SDK batches
 * uploads, so we flush asynchronously to make the UI snappy.
 */
import { Hono } from 'hono';
import { getLangfuse, flushLangfuse } from '../agent/langfuse.js';

export const feedbackRoute = new Hono();

type FeedbackBody = {
  trace_id?: string;
  observation_id?: string;
  value?: number;
  comment?: string;
  name?: string;
  household_id?: string;
  chat_id?: string;
  turn?: number;
};

feedbackRoute.post('/', async (c) => {
  const body = await c.req.json<FeedbackBody>().catch(() => ({}) as FeedbackBody);

  if (!body.trace_id) {
    return c.json({ ok: false, error: 'trace_id is required' }, 400);
  }
  if (typeof body.value !== 'number') {
    return c.json({ ok: false, error: 'value (number) is required' }, 400);
  }

  const lf = getLangfuse();
  if (!lf) {
    // Langfuse not configured — accept silently so the UI doesn't break.
    console.warn('[feedback] received but langfuse is disabled');
    return c.json({ ok: true, recorded: false });
  }

  try {
    lf.score({
      traceId: body.trace_id,
      observationId: body.observation_id,
      name: body.name ?? 'user-feedback',
      value: body.value,
      dataType: 'NUMERIC',
      comment: body.comment,
    });

    void flushLangfuse();
    return c.json({ ok: true, recorded: true });
  } catch (err) {
    console.error('[feedback] failed:', err);
    return c.json({ ok: false, error: (err as Error).message }, 500);
  }
});

File: backend_py/stackwealth/api/feedback.py
"""/api/feedback — user feedback → Langfuse score."""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..langfuse_client import flush_langfuse, get_langfuse

router = APIRouter()


@router.post("")
async def feedback(request: Request) -> JSONResponse:
    body: dict[str, Any] = await request.json() if (await request.body()) else {}

    if not body.get("trace_id"):
        return JSONResponse(content={"ok": False, "error": "trace_id is required"}, status_code=400)
    value = body.get("value")
    if not isinstance(value, (int, float)):
        return JSONResponse(content={"ok": False, "error": "value (number) is required"}, status_code=400)

    lf = get_langfuse()
    if lf is None:
        print("[feedback] received but langfuse is disabled")
        return JSONResponse(content={"ok": True, "recorded": False})

    try:
        lf.score(
            trace_id=body["trace_id"],
            observation_id=body.get("observation_id"),
            name=body.get("name") or "user-feedback",
            value=value,
            data_type="NUMERIC",
            comment=body.get("comment"),
        )
        flush_langfuse()
        return JSONResponse(content={"ok": True, "recorded": True})
    except Exception as err:
        print(f"[feedback] failed: {err}")
        return JSONResponse(content={"ok": False, "error": str(err)}, status_code=500)
